In [0]:
spark.sql("USE CATALOG capstone")
spark.sql("USE SCHEMA silver")

DataFrame[]

In [0]:
df_adm = spark.table("capstone.bronze.admissions")
df_pat = spark.table("capstone.bronze.patients")
df_doc = spark.table("capstone.bronze.doctors")
df_hosp = spark.table("capstone.bronze.hospitals")
df_lab = spark.table("capstone.bronze.lab_results")
df_bill = spark.table("capstone.bronze.billing")

In [0]:
from pyspark.sql.functions import col, row_number
from pyspark.sql.window import Window

In [0]:
#Check Missing Values
from pyspark.sql.functions import col, sum

def check_nulls(df, name):
    print(f"\nMissing values in {name}:")
    df.select([sum(col(c).isNull().cast("int")).alias(c) for c in df.columns]).show()

check_nulls(df_adm, "df_adm")
check_nulls(df_pat, "df_pat")
check_nulls(df_doc, "df_doc")
check_nulls(df_hosp, "df_hosp")
check_nulls(df_bill, "df_bill")
check_nulls(df_lab, "df_lab")



Missing values in df_adm:
+------------+----------+-----------+----------+--------------+--------------+--------------+--------------+-------------------+
|admission_id|patient_id|hospital_id|department|admission_date|discharge_date|admission_type|diagnosis_code|attending_doctor_id|
+------------+----------+-----------+----------+--------------+--------------+--------------+--------------+-------------------+
|        NULL|      NULL|       NULL|      NULL|          NULL|          NULL|          NULL|          NULL|               NULL|
+------------+----------+-----------+----------+--------------+--------------+--------------+--------------+-------------------+


Missing values in df_pat:
+----------+---+------+----+--------------+----------------------+
|patient_id|age|gender|city|insurance_type|chronic_condition_flag|
+----------+---+------+----+--------------+----------------------+
|         0|496|     0|   0|             0|                     0|
+----------+---+------+----+----

From above result we can find that we have 496 null values in df_pat[age] rest are fine

Handling missing values

In [0]:
from pyspark.sql.functions import col

median_age = df_pat.approxQuantile("age", [0.5], 0.01)[0]
print("Median Age:", median_age)

Median Age: 45.0


In [0]:
df_pat = df_pat.fillna({"age": median_age})

In [0]:
df_pat.filter(col("age").isNull()).count()

0

In [0]:
#Check Missing Values
from pyspark.sql.functions import col, sum

def check_nulls(df, name):
    print(f"\nMissing values in {name}:")
    df.select([sum(col(c).isNull().cast("int")).alias(c) for c in df.columns]).show()

check_nulls(df_adm, "df_adm")
check_nulls(df_pat, "df_pat")
check_nulls(df_doc, "df_doc")
check_nulls(df_hosp, "df_hosp")
check_nulls(df_bill, "df_bill")
check_nulls(df_lab, "df_lab")



Missing values in df_adm:
+------------+----------+-----------+----------+--------------+--------------+--------------+--------------+-------------------+
|admission_id|patient_id|hospital_id|department|admission_date|discharge_date|admission_type|diagnosis_code|attending_doctor_id|
+------------+----------+-----------+----------+--------------+--------------+--------------+--------------+-------------------+
|        NULL|      NULL|       NULL|      NULL|          NULL|          NULL|          NULL|          NULL|               NULL|
+------------+----------+-----------+----------+--------------+--------------+--------------+--------------+-------------------+


Missing values in df_pat:
+----------+---+------+----+--------------+----------------------+
|patient_id|age|gender|city|insurance_type|chronic_condition_flag|
+----------+---+------+----+--------------+----------------------+
|         0|  0|     0|   0|             0|                     0|
+----------+---+------+----+----

Duplicates

In [0]:
#Check Duplicates
from pyspark.sql.functions import col

def check_duplicates(df, name):
    print(f"\nDuplicates in {name}:")
    
    total_count = df.count()
    distinct_count = df.dropDuplicates().count()
    
    print(f"Total Rows     : {total_count}")
    print(f"Distinct Rows  : {distinct_count}")
    print(f"Duplicate Rows : {total_count - distinct_count}")

check_duplicates(df_adm, "df_adm")
check_duplicates(df_lab, "df_lab")
check_duplicates(df_pat, "df_pat")
check_duplicates(df_doc, "df_doc")
check_duplicates(df_hosp, "df_hosp")
check_duplicates(df_bill, "df_bill")



Duplicates in df_adm:
Total Rows     : 0
Distinct Rows  : 0
Duplicate Rows : 0

Duplicates in df_lab:
Total Rows     : 1000000
Distinct Rows  : 1000000
Duplicate Rows : 0

Duplicates in df_pat:
Total Rows     : 100000
Distinct Rows  : 100000
Duplicate Rows : 0

Duplicates in df_doc:
Total Rows     : 2000
Distinct Rows  : 2000
Duplicate Rows : 0

Duplicates in df_hosp:
Total Rows     : 50
Distinct Rows  : 50
Duplicate Rows : 0

Duplicates in df_bill:
Total Rows     : 500000
Distinct Rows  : 500000
Duplicate Rows : 0


In [0]:
df_adm = df_adm.dropDuplicates()
df_adm.count()

0

In [0]:
check_duplicates(df_adm, "df_adm_after_cleaning")


Duplicates in df_adm_after_cleaning:
Total Rows     : 0
Distinct Rows  : 0
Duplicate Rows : 0


In [0]:
#Check Duplicates
from pyspark.sql.functions import col

def check_duplicates(df, name):
    print(f"\nDuplicates in {name}:")
    
    total_count = df.count()
    distinct_count = df.dropDuplicates().count()
    
    print(f"Total Rows     : {total_count}")
    print(f"Distinct Rows  : {distinct_count}")
    print(f"Duplicate Rows : {total_count - distinct_count}")

check_duplicates(df_adm, "df_adm")
check_duplicates(df_lab, "df_lab")
check_duplicates(df_pat, "df_pat")
check_duplicates(df_doc, "df_doc")
check_duplicates(df_hosp, "df_hosp")
check_duplicates(df_bill, "df_bill")



Duplicates in df_adm:
Total Rows     : 0
Distinct Rows  : 0
Duplicate Rows : 0

Duplicates in df_lab:
Total Rows     : 1000000
Distinct Rows  : 1000000
Duplicate Rows : 0

Duplicates in df_pat:
Total Rows     : 100000
Distinct Rows  : 100000
Duplicate Rows : 0

Duplicates in df_doc:
Total Rows     : 2000
Distinct Rows  : 2000
Duplicate Rows : 0

Duplicates in df_hosp:
Total Rows     : 50
Distinct Rows  : 50
Duplicate Rows : 0

Duplicates in df_bill:
Total Rows     : 500000
Distinct Rows  : 500000
Duplicate Rows : 0


Logical Duplicates

In [0]:
#Logical duplicates(Same patient same day)
from pyspark.sql.functions import col

df_duplicates = df_adm.groupBy("patient_id", "admission_date") \
                     .count() \
                     .filter(col("count") > 1)

df_duplicates.show()


+----------+--------------+-----+
|patient_id|admission_date|count|
+----------+--------------+-----+
+----------+--------------+-----+



In [0]:
#Same patient same day same hospital
df_adm.groupBy("patient_id", "hospital_id", "admission_date").count().filter("count > 1").count()

0

Primary key tests

In [0]:

df_pat.groupBy("patient_id").count().filter("count > 1").count()


0

In [0]:
df_doc.groupBy("doctor_id").count().filter("count > 1").count()

0

In [0]:
df_hosp.groupBy("hospital_id").count().filter("count > 1").count()

0

In [0]:
df_lab.groupBy("lab_test_id").count().filter("count > 1").count()

0

In [0]:
df_lab_duplicates = df_lab.join(
    df_lab.groupBy("patient_id", "test_name", "test_date")
          .count()
          .filter(col("count") > 1),
    ["patient_id", "test_name", "test_date"]
).orderBy("patient_id", "test_name", "test_date")

df_lab_duplicates.show()

+----------+----------+----------+-----------+-----------+-----------------+-------------------+-----+
|patient_id| test_name| test_date|lab_test_id|test_result|normal_range_flag|        reported_at|count|
+----------+----------+----------+-----------+-----------+-----------------+-------------------+-----+
|     10029|     X-Ray|2024-03-02|    1178293|      81.44|                0|2024-03-03 11:00:00|    2|
|     10029|     X-Ray|2024-03-02|    1704835|      37.19|                0|2024-03-02 03:00:00|    2|
|     10137|Blood Test|2023-11-29|    1002387|      22.22|                0|2023-11-30 13:00:00|    2|
|     10137|Blood Test|2023-11-29|    1529424|       85.2|                0|2023-11-29 08:00:00|    2|
|     10215|   CT Scan|2023-01-23|    1151638|      35.69|                0|2023-01-24 18:00:00|    2|
|     10215|   CT Scan|2023-01-23|    1171315|      49.01|                0|2023-01-23 13:00:00|    2|
|     10241|Urine Test|2023-12-03|    1300863|      15.38|               

In [0]:
from pyspark.sql.window import Window
from pyspark.sql.functions import row_number, col

window_spec = Window.partitionBy("patient_id", "test_name", "test_date") \
                    .orderBy(col("reported_at").desc())

In [0]:
df_lab_clean = df_lab.withColumn("rn", row_number().over(window_spec)) \
                     .filter(col("rn") == 1) \
                     .drop("rn")

In [0]:
df_lab_clean.groupBy("patient_id", "test_name", "test_date") \
            .count() \
            .filter(col("count") > 1) \
            .show()

+----------+---------+---------+-----+
|patient_id|test_name|test_date|count|
+----------+---------+---------+-----+
+----------+---------+---------+-----+



Billing Claims

In [0]:
df_bill.groupBy("bill_id").count().filter("count > 1").count()

0

In [0]:
#one admission usually has one bill
df_bill.groupBy("admission_id").count().filter("count > 1").count()

131947

In [0]:
from pyspark.sql.functions import col

df_dup_keys = df_bill.groupBy("admission_id") \
                    .count() \
                    .filter(col("count") > 1)

df_bill_duplicates = df_bill.join(
    df_dup_keys,
    on="admission_id",
    how="inner"
).orderBy("admission_id")

df_bill_duplicates.show(10)

+------------+-------+-------------+------------------------+------------------+------------+-----+
|admission_id|bill_id|total_charges|insurance_covered_amount|   patient_payable|claim_status|count|
+------------+-------+-------------+------------------------+------------------+------------+-----+
|      700001| 994911|    143291.86|                70081.79| 73210.06999999999|    Rejected|    2|
|      700001| 637758|     66343.58|                57666.72|           8676.86|    Rejected|    2|
|      700004| 819433|     19010.64|                15327.66|3682.9799999999996|    Approved|    2|
|      700004| 795860|     61863.31|                40136.64|          21726.67|    Approved|    2|
|      700005| 899932|    118300.94|                59796.58|          58504.36|     Pending|    2|
|      700005| 680667|    122359.58|               100343.93| 22015.65000000001|    Approved|    2|
|      700007| 906879|     53906.37|                19389.84|          34516.53|    Rejected|    3|


In [0]:
from pyspark.sql.window import Window
from pyspark.sql.functions import row_number, col

window_spec = Window.partitionBy("admission_id") \
                    .orderBy(col("bill_id").desc())

df_bill_clean = df_bill.withColumn("rn", row_number().over(window_spec)) \
                       .filter(col("rn") == 1) \
                       .drop("rn")

In [0]:
df_bill_clean.groupBy("admission_id") \
             .count() \
             .filter("count > 1") \
             .count()

0

Derivation of columns

In [0]:
from pyspark.sql.functions import datediff

df_adm = df_adm.withColumn(
    "LOS",
    datediff(col("discharge_date"), col("admission_date"))
)

In [0]:
df_adm = df_adm.withColumn(
    "risk_flag",
    (col("LOS") > 10).cast("int")
)

In [0]:
df_adm.display()

admission_id,patient_id,hospital_id,department,admission_date,discharge_date,admission_type,diagnosis_code,attending_doctor_id,LOS,risk_flag


Re admission flag

In [0]:
from pyspark.sql.window import Window
from pyspark.sql.functions import col, lead, datediff

window_spec = Window.partitionBy("patient_id") \
                    .orderBy("admission_date")

In [0]:
df_adm = df_adm.withColumn(
    "next_admission_date",
    lead("admission_date").over(window_spec)
)

In [0]:
df_adm = df_adm.withColumn(
    "days_to_next_admission",
    datediff(col("next_admission_date"), col("discharge_date"))
)

In [0]:
from pyspark.sql.functions import when

df_adm = df_adm.withColumn(
    "readmission_flag",
    when((col("days_to_next_admission") <= 30) & (col("days_to_next_admission") >= 0), 1).otherwise(0)
)

In [0]:
df_adm = df_adm.drop("next_admission_date", "days_to_next_admission")

In [0]:
df_adm.select("patient_id", "admission_date", "discharge_date", "readmission_flag").display()

patient_id,admission_date,discharge_date,readmission_flag


In [0]:
df_adm.write.format("delta") \
.mode("overwrite") \
.partitionBy("admission_date", "hospital_id") \
.option("overwriteSchema", "true").saveAsTable("capstone.silver.admissions")

In [0]:
df_adm.write.mode("overwrite").option("mergeSchema", "true").saveAsTable("capstone.silver.admissions")
df_pat.write.mode("overwrite").option("mergeSchema", "true").saveAsTable("capstone.silver.patients")
df_doc.write.mode("overwrite").option("mergeSchema", "true").saveAsTable("capstone.silver.doctors")
df_hosp.write.mode("overwrite").option("mergeSchema", "true").saveAsTable("capstone.silver.hospitals")
df_bill.write.mode("overwrite").option("mergeSchema", "true").saveAsTable("capstone.silver.billing")
df_lab.write.mode("overwrite").option("mergeSchema", "true").saveAsTable("capstone.silver.lab_results")

In [0]:
spark.sql("DESCRIBE DETAIL capstone.silver.admissions").show(truncate=False)

+------+------------------------------------+--------------------------+-----------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-----------------------+-------------------+-----------------------------+-----------------+--------+-----------+------------------------------------------------------------------------------+----------------+----------------+-----------------------------------------+---------------------------------------------------------------+-------------+
|format|id                                  |name                      |description|location                                                                                                                                                                                          |createdAt              |lastModified       |partitionColumns             |clusteringColumns|num